File: src/services/LoanService.cpp
#include "services/LoanService.h"

#include <cmath>

#include "core/Utils.h"

namespace bms {

long long LoanService::computeEmi(long long principalPaise,
                                  double ratePerAnnum,
                                  int tenureMonths) {
    if (principalPaise <= 0 || tenureMonths <= 0) return 0;
    const long double p = static_cast<long double>(principalPaise);
    const long double mr = ratePerAnnum / 100.0L / 12.0L;
    if (mr == 0.0L) {
        return std::llround(p / static_cast<long double>(tenureMonths));
    }
    const long double factor = std::pow(1.0L + mr, tenureMonths);
    const long double emi = p * mr * factor / (factor - 1.0L);
    return std::llround(emi);
}

Loan* LoanService::applyLoan(int customerId, const std::string& accountNumber,
                             long long principalPaise, double ratePerAnnum,
                             int tenureMonths, std::string& err) {
    if (db_.findCustomerById(customerId) == nullptr) {
        err = "No such customer.";
        return nullptr;
    }
    const Account* acc = db_.findAccountByNumber(accountNumber);
    if (!acc || acc->customerId != customerId) {
        err = "Loan account does not belong to this customer.";
        return nullptr;
    }
    if (acc->status != AccountStatus::ACTIVE) {
        err = "Loan account is not active.";
        return nullptr;
    }
    if (principalPaise <= 0) {
        err = "Loan amount must be positive.";
        return nullptr;
    }
    if (ratePerAnnum < 0.0 || ratePerAnnum > 36.0) {
        err = "Rate must be between 0% and 36%.";
        return nullptr;
    }
    if (tenureMonths < 1 || tenureMonths > 120) {
        err = "Tenure must be between 1 and 120 months.";
        return nullptr;
    }
    if (findOpenLoanByAccount(accountNumber) != nullptr) {
        err = "An open loan already exists on this account.";
        return nullptr;
    }

    Loan loan;
    loan.loanId = db_.nextLoanId();
    loan.customerId = customerId;
    loan.accountNumber = accountNumber;
    loan.principalPaise = principalPaise;
    loan.outstandingPaise = principalPaise;
    loan.ratePerAnnum = ratePerAnnum;
    loan.tenureMonths = tenureMonths;
    loan.emiPaise = computeEmi(principalPaise, ratePerAnnum, tenureMonths);
    loan.emiPaidCount = 0;
    loan.amountPaidPaise = 0;
    loan.status = LoanStatus::PENDING;
    loan.appliedDate = Utils::now();
    loan.decisionDate = "";
    loan.nextDueDate = "";
    db_.loans().push_back(loan);
    db_.saveAll();
    return &db_.loans().back();
}

bool LoanService::approveLoan(const std::string& loanId, std::string& err) {
    Loan* loan = db_.findLoanById(loanId);
    if (!loan) {
        err = "Loan not found.";
        return false;
    }
    if (loan->status != LoanStatus::PENDING) {
        err = "Loan is not pending approval.";
        return false;
    }

    Account* acc = db_.findAccountByNumber(loan->accountNumber);
    if (!acc || acc->status != AccountStatus::ACTIVE) {
        err = "Linked account is not active.";
        return false;
    }

    loan->status = LoanStatus::ACTIVE;
    loan->emiPaise = computeEmi(loan->principalPaise, loan->ratePerAnnum,
                                loan->tenureMonths);
    loan->decisionDate = Utils::now();
    loan->nextDueDate = Utils::addMonths(Utils::today(), 1);

    acc->balancePaise += loan->principalPaise;

    Transaction tx;
    tx.id = db_.nextTransactionId();
    tx.accountNumber = acc->accountNumber;
    tx.type = TxType::LOAN_DISBURSAL;
    tx.amountPaise = loan->principalPaise;
    tx.balanceAfterPaise = acc->balancePaise;
    tx.counterparty = loan->loanId;
    tx.timestamp = Utils::now();
    tx.remark = "Loan disbursal — " + loan->loanId;
    db_.addTransaction(tx);

    db_.saveAll();
    return true;
}

bool LoanService::rejectLoan(const std::string& loanId, std::string& err) {
    Loan* loan = db_.findLoanById(loanId);
    if (!loan) {
        err = "Loan not found.";
        return false;
    }
    if (loan->status != LoanStatus::PENDING) {
        err = "Loan is not pending.";
        return false;
    }
    loan->status = LoanStatus::REJECTED;
    loan->decisionDate = Utils::now();
    db_.saveAll();
    return true;
}

bool LoanService::payEmi(const std::string& loanId, std::string& err) {
    Loan* loan = db_.findLoanById(loanId);
    if (!loan) {
        err = "Loan not found.";
        return false;
    }
    if (loan->status != LoanStatus::ACTIVE) {
        err = "Loan is not active.";
        return false;
    }

    Account* acc = db_.findAccountByNumber(loan->accountNumber);
    if (!acc || acc->status != AccountStatus::ACTIVE) {
        err = "Linked account is not active.";
        return false;
    }

    const long long due = loan->outstandingPaise < loan->emiPaise
                              ? loan->outstandingPaise
                              : loan->emiPaise;
    if (acc->balancePaise < due) {
        err = "Insufficient balance for this EMI.";
        return false;
    }

    acc->balancePaise -= due;
    loan->amountPaidPaise += due;
    loan->outstandingPaise -= due;
    loan->emiPaidCount += 1;

    if (loan->outstandingPaise <= 0) {
        loan->status = LoanStatus::CLOSED;
        loan->nextDueDate = "";
    } else {
        loan->nextDueDate = Utils::addMonths(loan->nextDueDate, 1);
    }

    Transaction tx;
    tx.id = db_.nextTransactionId();
    tx.accountNumber = acc->accountNumber;
    tx.type = TxType::EMI_PAYMENT;
    tx.amountPaise = due;
    tx.balanceAfterPaise = acc->balancePaise;
    tx.counterparty = loan->loanId;
    tx.timestamp = Utils::now();
    tx.remark = "EMI " + std::to_string(loan->emiPaidCount) + "/" +
                std::to_string(loan->tenureMonths) + " — " + loan->loanId;
    db_.addTransaction(tx);

    db_.saveAll();
    return true;
}

Loan* LoanService::findOpenLoanByAccount(const std::string& accountNumber) {
    for (auto& loan : db_.loans()) {
        if (loan.accountNumber == accountNumber &&
            (loan.status == LoanStatus::ACTIVE ||
             loan.status == LoanStatus::PENDING)) {
            return &loan;
        }
    }
    return nullptr;
}

}  // namespace bms

File: web/app/services/loan_service.py
"""Loan application, approval, EMI computation and repayment."""
import math

from .. import money
from ..database import get_db, next_id
from .account_service import get_account
from . import transaction_service


def compute_emi(principal_paise, rate, months) -> int:
    """Standard amortising EMI in paise."""
    principal = principal_paise / 100.0
    r = rate / 100.0 / 12.0
    if r == 0:
        return round(principal / months)
    emi = principal * r * (1 + r) ** months / ((1 + r) ** months - 1)
    return round(emi * 100)


def apply_loan(user_id, account_number, principal_paise, rate, months):
    """Submit a loan application. Returns (loan_id, None) or (None, error)."""
    if principal_paise <= 0:
        return None, "Loan amount must be positive."
    if not (1 <= rate <= 36):
        return None, "Interest rate must be between 1% and 36%."
    if not (3 <= months <= 120):
        return None, "Tenure must be between 3 and 120 months."

    db = get_db()
    acct = get_account(account_number)
    if acct is None or acct["user_id"] != user_id:
        return None, "Account not found or not yours."
    if acct["status"] != "Active":
        return None, "Account is closed — cannot apply for a loan."
    if acct["type"] == "Checking":
        return None, "Loans can only be linked to savings accounts."

    emi = compute_emi(principal_paise, rate, months)
    loan_id = f"LN{next_id(db, 'loan_seq'):04d}"
    db.execute(
        "INSERT INTO loans (loan_id, user_id, account_number, principal_paise,"
        " outstanding_paise, rate, tenure_months, emi_paise, status, applied_date)"
        " VALUES (?, ?, ?, ?, ?, ?, ?, ?, 'Pending', ?)",
        (loan_id, user_id, account_number, principal_paise, principal_paise,
         rate, months, emi, money.now()),
    )
    db.commit()
    return loan_id, None


def loans_of(user_id):
    db = get_db()
    return db.execute(
        "SELECT l.*, a.type AS account_type FROM loans l"
        " JOIN accounts a ON a.account_number = l.account_number"
        " WHERE l.user_id = ? ORDER BY l.applied_date DESC",
        (user_id,),
    ).fetchall()


def pending_loans():
    db = get_db()
    return db.execute(
        "SELECT l.*, u.name AS user_name, a.type AS account_type FROM loans l"
        " JOIN users u ON u.id = l.user_id"
        " JOIN accounts a ON a.account_number = l.account_number"
        " WHERE l.status = 'Pending' ORDER BY l.applied_date",
    ).fetchall()


def active_loans():
    db = get_db()
    return db.execute(
        "SELECT l.*, u.name AS user_name, a.type AS account_type FROM loans l"
        " JOIN users u ON u.id = l.user_id"
        " JOIN accounts a ON a.account_number = l.account_number"
        " WHERE l.status = 'Active' ORDER BY l.applied_date",
    ).fetchall()


def get_loan(loan_id):
    return get_db().execute("SELECT * FROM loans WHERE loan_id = ?", (loan_id,)).fetchone()


def approve_loan(loan_id):
    """Approve a pending loan and disburse to the linked account."""
    db = get_db()
    loan = get_loan(loan_id)
    if loan is None:
        return "Loan not found."
    if loan["status"] != "Pending":
        return "Loan is not pending approval."
    acct = get_account(loan["account_number"])
    if acct is None or acct["status"] != "Active":
        return "Linked account is closed."

    due_date = money.add_months(money.today(), 1)
    db.execute(
        "UPDATE loans SET status = 'Active', decision_date = ?, next_due_date = ?"
        " WHERE loan_id = ?",
        (money.now(), due_date, loan_id),
    )
    new_balance = acct["balance_paise"] + loan["principal_paise"]
    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_balance, loan["account_number"]),
    )
    transaction_service._record(
        db, loan["account_number"], "DEPOSIT", loan["principal_paise"], new_balance,
        "OneSky Bank", f"Loan disbursement {loan_id}",
    )
    db.commit()
    return None


def reject_loan(loan_id):
    db = get_db()
    loan = get_loan(loan_id)
    if loan is None:
        return "Loan not found."
    if loan["status"] != "Pending":
        return "Loan is not pending approval."
    db.execute(
        "UPDATE loans SET status = 'Rejected', decision_date = ? WHERE loan_id = ?",
        (money.now(), loan_id),
    )
    db.commit()
    return None


def pay_emi(loan_id):
    """Pay one EMI from the linked account. Returns error or None."""
    db = get_db()
    loan = get_loan(loan_id)
    if loan is None:
        return "Loan not found."
    if loan["status"] != "Active":
        return "Loan is not active."
    acct = get_account(loan["account_number"])
    if acct is None or acct["status"] != "Active":
        return "Linked account is closed."
    if loan["outstanding_paise"] <= 0:
        return "Loan is already fully repaid."

    due = min(loan["emi_paise"], loan["outstanding_paise"])
    if due > acct["balance_paise"]:
        return f"Insufficient balance to pay EMI. Available: {money.format_paise(acct['balance_paise'])}"

    new_balance = acct["balance_paise"] - due
    outstanding = loan["outstanding_paise"] - due
    amount_paid = loan["amount_paid_paise"] + due
    emi_count = loan["emi_paid_count"] + 1

    db.execute(
        "UPDATE accounts SET balance_paise = ? WHERE account_number = ?",
        (new_balance, loan["account_number"]),
    )
    next_due = loan["next_due_date"]
    if outstanding <= 0:
        db.execute(
            "UPDATE loans SET outstanding_paise = 0, amount_paid_paise = ?,"
            " emi_paid_count = ?, status = 'Closed', next_due_date = '' WHERE loan_id = ?",
            (amount_paid, emi_count, loan_id),
        )
    else:
        next_due = money.add_months(next_due, 1)
        db.execute(
            "UPDATE loans SET outstanding_paise = ?, amount_paid_paise = ?,"
            " emi_paid_count = ?, next_due_date = ? WHERE loan_id = ?",
            (outstanding, amount_paid, emi_count, next_due, loan_id),
        )
    transaction_service._record(
        db, loan["account_number"], "WITHDRAW", due, new_balance,
        "OneSky Bank", f"EMI payment for {loan_id}",
    )
    db.commit()
    return None
